### PR TITLE
Manage a stack of local graphic state

### PR DESCRIPTION
**Pull request by gmischler**
as of #250 (second part).

* New ability to change graphics state settings temporarily.
* New context manager .local_context() to create a temporary state environment.
* Rotations don't conflict with graphics state settings anymore.
* Templates don't need to jump through hoops anymore to rotate stuff.

### CODE REVIEW

File: fpdf/graphics_state.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                draw_color="0 G",
+                fill_color="0 g",
+                text_color="0 g",
+                underline=False,
+                font_style="",
+                font_stretching=100,
+                font_family="",
+                font_size_pt=0,
+                font_size=0,
+                dash_pattern="[] 0 d",
```
Comment (Lucas-C) [inline fpdf/graphics_state.py:25]: Note `line_width`?

File: fpdf/graphics_state.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        self.__statestack.append(self.__statestack[-1].copy())
+
+    def _pop_local_stack(self):
+        del self.__statestack[-1]
+
+    @property
+    def draw_color(self):
+        return self.__statestack[-1]["draw_color"]
+
+    @draw_color.setter
```
Comment (Lucas-C) [inline fpdf/graphics_state.py:39]: Il faut que les "setters" de la classe parente fassent désormais appel à ces méthodes : `FPDF.set_stretching`, `FPDF.set_draw_color`, `FPDF.set_fill_color`, `FPDF.set_text_color`, `FPDF.set_line_width`, `FPDF.set_dash_pattern`, `FPDF.set_font`, `FPDF.set_font_size`

File: test/template/test_flextemplate.py [hunk 3]
```diff
@@ -192,7 +192,6 @@ def test_flextemplate_rotation(tmp_path):
             "type": "C39",
             "x1": 80,
             "y1": 10,
-            "x2": 70,
```
Comment (Lucas-C) [inline test/template/test_flextemplate.py:195]: Why changing those values?
It'b better if the exeact same PDF file could is generated BEFORE & AFTER this PR, to ensure backward-compatibility

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Good job overall @gmischler !

I made a few comments on your PR

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,253 +1,255 @@
-Changelog
```
[the remark this answers is not in the document]
Comment (gmischler) [reply CHANGELOG.md:1]: Looks like it. My local file is formatted exactly as the the other's, though...

File: fpdf/fpdf.py [hunk 5]
```diff
@@ -246,6 +247,7 @@ def __init__(
                 `None` disables font chaching.
                 The default is `True`, meaning the current folder.
         """
+        GraphicsStateMixin.__init__(self)
```
[the remark this answers is not in the document]
Comment (Lucas-C) [reply fpdf/fpdf.py:250]: > there's no way around naming them individually
I'm not so sure. Actually, this works pretty well:
```python
class AMixin:
    def __init__(self):
        super().__init__()
        print("AMixin.__init__")

class BMixin:
    def __init__(self):
        super().__init__()
        print("BMixin.__init__")

class FPDF(AMixin, BMixin):
    def __init__(self):
        super().__init__()
        print("FPDF.__init__")

class MyPDF(FPDF):
    pass

print(MyPDF())
```

File: fpdf/graphics_state.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    normal instance attributes. But by the magic of properties, we can
+    push and pop levels as needed, and users will always see and modify
+    just the current version.
+
+    This class is mixed in by fpdf.FPDF(), and is not meant to be used
+    directly by user code.
+    """
+
+    def __init__(self):
+        self.__statestack = [
```
[the remark this answers is not in the document]
Comment (Lucas-C) [reply fpdf/graphics_state.py:14]: Alright

File: fpdf/graphics_state.py (hunk 2 above)
Comment (Lucas-C) [reply fpdf/graphics_state.py:40]: (sorry for the sudden mix of French in the review, my mistake)

File: test/template/test_flextemplate.py (hunk 3 above)
Comment (Lucas-C) [reply test/template/test_flextemplate.py:195]: Ok then, good job

File: fpdf/fpdf.py [hunk 7]
```diff
@@ -263,16 +265,19 @@ def __init__(
...
         self.current_font = {}  # current font
+        self.str_alias_nb_pages = "{nb}"
+        # graphics state variables from the stack
         self.font_family = ""  # current font family
         self.font_style = ""  # current font style
         self.font_size_pt = 12  # current font size in points
         self.font_stretching = 100  # current font stretching
-        self.str_alias_nb_pages = "{nb}"
         self.underline = 0  # underlining flag
         self.draw_color = "0 G"
```
[the remark this answers is not in the document]
Comment (Lucas-C) [reply fpdf/fpdf.py:277]: Alright, I understand now.

Could you please extend your "graphics state variables from the stack" comment here, in `FPDF.__init__`,
to make explicit that we are **not** initializing properties at this point,
but setting default values of properties defined in `GraphicsStateMixin`?

File: fpdf/graphics_state.py (hunk 2 above)
Comment (Lucas-C) [reply fpdf/graphics_state.py:40]: Ok, I see now why this is not needed

File: fpdf/fpdf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmischler) [reply fpdf/fpdf.py:250]: So you're saying I had the wrong sequence in my head all those years?
Go figure...

(I used to write tons of code for Python 1.5, so I'll use that as my excuse... :wink:)

File: fpdf/fpdf.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gmischler) [reply fpdf/fpdf.py:277]: We don't seem to think of the same operation when saying "initializing properties".
I'll find a phrasing that makes it clear what happens either way.

[the remark this answers is not in the document]
Comment (Lucas-C) [reply fpdf/fpdf.py]: That's perfect, thank you!

**Review by Lucas-C — APPROVED** [review]

**Comment by Lucas-C** [comment]
Thank you @gmischler for this great code refactoring!

**Comment by Lucas-C** [comment]
FYI I mentioned you in this blog article 😉
https://chezsoi.org/lucas/blog/hacktoberfest-on-fpdf2.html

**Pull request merged**